kano_profile/tracker/tracking_sessions.py

```diff
@@ -0,0 +1,315 @@
+
+    return data.get('app_session_id', "")
+
+
+def session_start(name, pid=None, ignore_pause=False):
+    if not pid:
+        pid = os.getpid()
+    pid = int(pid)
+
+    if not ignore_pause and is_tracking_paused():
+        session = TrackingSession(name=name, pid=pid)
+        try:
+            paused_sessions_f = open_locked(PAUSED_SESSIONS_FILE, 'a')
+        except IOError as err:
+            logger.error(
+                'Error while opening the paused sessions file: {}'.format(err)
+            )
+        else:
+            paused_sessions_f.write(
+                '{}\n'.format(session.dumps())
+            )
+
+            return session.path
+
+    data = {
+        'pid': pid,
+        'name': name,
+        'started': int(time.time()),
+        'elapsed': 0,
+        'app_session_id': str(uuid5(uuid1(), name + str(pid))),
+        'finished': False,
+        'token-system': TOKEN
+    }
+
+    path = get_session_file_path(data['name'], data['pid'])
+
+    try:
+        f = open_locked(path, 'w')
+    except IOError as e:
+        logger.error("Error opening tracker session file {}".format(e))
+    else:
+        with f:
+            json.dump(data, f)
+        if 'SUDO_USER' in os.environ:
+            chown_path(path)
+
+    return path
+
+
+def session_end(session_file):
+    if not os.path.exists(session_file):
+        msg = "Someone removed the tracker file, the runtime of this " \
+              "app will not be logged"
+        logger.warn(msg)
+        return
+
+    try:
+        rf = open_locked(session_file, 'r')
+    except IOError as e:
+        logger.error("Error opening the tracker session file {}".format(e))
+    else:
+        with rf:
+            data = json.load(rf)
+
+            data['elapsed'] = int(time.time()) - data['started']
+            data['finished'] = True
+
+            try:
+                wf = open(session_file, 'w')
+            except IOError as e:
+                logger.error(
+                    "Error opening the tracker session file {}".format(e))
+            else:
+                with wf:
+                    json.dump(data, wf)
+        if 'SUDO_USER' in os.environ:
+            chown_path(session_file)
+
+
+def get_paused_sessions():
+    if not os.path.exists(PAUSED_SESSIONS_FILE):
+        return []
+
+    try:
+        sessions_f = open_locked(PAUSED_SESSIONS_FILE, 'r')
+    except IOError as err:
+        logger.error('Error opening the paused sessions file: {}'.format(err))
+        return []
+    else:
+        with sessions_f:
+            paused_sessions = []
+            for session in sessions_f:
+                if not session:
+                    continue
+
+                paused_sessions.append(
+                    TrackingSession.loads(session)
+                )
+
+            return paused_sessions
+
+
+def is_tracking_paused():
+    return os.path.exists(PAUSED_SESSIONS_FILE)
+
+
+def pause_tracking_session(session):
```

Comment:
Methods like this one - `pause`, `start`, `stop` to me, belong in the `TrackingSession` class and ones that manage all of them in some sort of `Controller`. This is more of a general note about how things are structured here.

Reply:
I agree with this. It is the reason that I added the `TrackingSession` class but for fear of breaking more things (and getting stuff done) settled with leaving the `start` and `stop` here and introduced similar interfaces for the `pause` functions but definitely pro this moving forward.